pandas/io/json/_json.py

```diff
@@ -33,7 +35,7 @@
 # interface to/from
 def to_json(
     path_or_buf,
-    obj,
+    obj: NDFrame,
```

Comment:
should this be FrameOrSeries?

Reply:
The problem with ``FrameOrSeries`` is that it does not allow changing the type of return value.
Inside the function ``to_json`` there is a statement, which changes object from series to frame, which is not allowed.
```
    if orient == "table" and isinstance(obj, Series):
        obj = obj.to_frame(name=obj.name or "values")
```

mypy error:
```
pandas/io/json/_json.py:78: error: Incompatible types in assignment (expression has type "DataFrame", variable has type "FrameOrSeries")  [assignment]
```

I tried to use ``FrameOrSeriesUnion``, which allows changing type from series to frame, but I get another mypy error.
Apparently for ``NDFrame`` only ``FrameOrSeries`` is a correct casting.

```
pandas/core/generic.py:2159: error: The erased type of self "Union[pandas.core.frame.DataFrame, pandas.core.series.Series]" is not a supertype of its class "pandas.core.generic.NDFrame"  [misc]
```

So, I have no clue how to deal with typing here.

Reply:
@willayd i expected FrameOrSeriesUnion would work for this, am i missing something?

Reply:
@ivanovmg can you merge master and update the error message here? I don't think that line number is reflective of the latest code base

I think the error message is stemming from a mixture of using a Union and a TypeVar (the latter being somewhere in generic). Ideally should be cleaned up but I don't think needs to hold up this PR

Reply:
Here we go.
I merged master, then modified ``pandas/core/generic.py``.

line 2164:
```
    def to_json(
        self: FrameOrSeriesUnion,
        path_or_buf: Optional[FilePathOrBuffer] = None,
```

And replaced ``NDFrame`` with ``FrameOrSeriesUnion``.

The error would be:
```
pandas/core/generic.py:2164: error: The erased type of self "Union[pandas.core.frame.DataFrame, pandas.core.series.Series]" is not a supertype of its class "pandas.core.generic.NDFrame"  [misc]
```

However, if I do not touch ``generic.py``, then the following error with be thrown.
```
pandas/core/generic.py:2437: error: Argument "obj" to "to_json" has incompatible type "NDFrame"; expected "Union[DataFrame, Series]"  [arg-type]
```